### PR TITLE
Add support for jupyter notebook

### PR DESCRIPTION
**Pull request by Fernando-crz**
- Commit rebased from https://github.com/bndr/pipreqs/pull/210 . 
- Credits to pakio: https://github.com/pakio

### CODE REVIEW

**Comment by alan-barzilay** [comment]
tests failed due to nbconvert version, have you run the tests locally?
Also, have you addressed the issue raised by @jamesmyatt about optional dependency?

**Comment by jamesmyatt** [comment]
I think @alan-barzilay means this one: https://github.com/bndr/pipreqs/pull/210#discussion_r734564597

**Comment by Fernando-crz** [comment]
Just fixed bugs related to the tests. The tests were failing because they were not considering that ipython was now part of the locally installed packages. Just fixed this issue now and tests are running just fine, but for some odd reason the codecov coverage report failed. @alan-barzilay

**Comment by Fernando-crz** [comment]
Just added support for jupyter notebook as an optional depency, as adressed by https://github.com/bndr/pipreqs/pull/210#discussion_r734564597.
@alan-barzilay

File: pipreqs/pipreqs.py [hunk 1]
```diff
@@ -473,20 +513,21 @@ def init(args):
...
-                      x.lower() not in [y for x in local for y in x['exports']]
-                      and
-                      # check if candidate is package names
-                      x.lower() not in [x['name'] for x in local]]
-
-        imports = local + get_imports_info(difference,
-                                           proxy=proxy,
-                                           pypi_server=pypi_server)
+        difference = [
+            x
```
Comment (alan-barzilay) [inline pipreqs/pipreqs.py:517]: formatters are weird

File: setup.py [hunk 2]
```diff
@@ -15,7 +15,11 @@
     history = history_file.read().replace('.. :changelog:', '')
 
 requirements = [
-    'docopt', 'yarg', 'nbconvert', 'ipython'
+    'docopt', 'yarg'
+]
+
+jupyter_requirements = [
```
Comment (alan-barzilay) [inline setup.py:21]: please eli5 how this works or link a reference haha

File: setup.py [hunk 3]
```diff
@@ -34,6 +38,7 @@
     include_package_data=True,
     package_data={'': ['stdlib', 'mapping']},
     install_requires=requirements,
+    extras_require={"jupyter":jupyter_requirements},
```
Comment (alan-barzilay) [inline setup.py:41]: woof woof

**Review by alan-barzilay — CHANGES_REQUESTED** [review]
guys, you gotta pay closer attention when you are using someone else's commit haha  even more so when they are that old and things have diverged quite a bii. Its ok to pass them through a formatter or to create a fixup  to alter part of their code (i believe that would give both of you authorship of the commit, and thats ok)

File: pipreqs/pipreqs.py [hunk 4]
```diff
@@ -47,17 +47,18 @@
 import requests
 from yarg import json2package
 from yarg.exceptions import HTTPError
+from nbconvert import PythonExporter
 
 from pipreqs import __version__
 
 REGEXP = [
-    re.compile(r'^import (.+)$'),
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:54]: Since the repository is using double quotes as standard, i will just leave it as it is

File: pipreqs/pipreqs.py [hunk 5]
```diff
@@ -137,70 +158,93 @@ def get_all_imports(
...
+
+    Returns:
+        str: parsed string
+
+    """
+
+    exporter = PythonExporter()
+    (body, _) = exporter.from_filename(file_name)
+
+    return body.encode(encoding if encoding is not None else "utf-8")
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:196]: done

File: pipreqs/pipreqs.py [hunk 6]
```diff
@@ -137,70 +158,93 @@ def get_all_imports(
...
-    logging.debug('Found packages: {0}'.format(packages))
+    logging.debug("Found packages: {0}".format(packages))
 
     with open(join("stdlib"), "r") as f:
         data = {x.strip() for x in f}
 
     return list(packages - data)
 
 
+def filter_line(line):
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: Done

File: pipreqs/pipreqs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:192]: Setting the default value to utf-8 makes the tests fail. It happens because, when calling the `init` function without an encoding option, it defaults to _None_, passing _None_ as an argument to to the `get_all_imports` function, which passes _None_ as the encoding argument to the `ipynb_2_py` function. Leaving the function working in this way, makes it capable of dealing with _None_ values as input without crashing the program.

File: pipreqs/pipreqs.py [hunk 7]
```diff
@@ -256,14 +298,14 @@ def get_import_local(imports, encoding=None):
...
-            if item in package['exports'] or item == package['name']:
+            if item in package["exports"] or item == package["name"]:
                 result.append(package)
 
     # removing duplicates of package/version
     # had to use second method instead of the previous one,
     # because we have a list in the 'exports' field
     # https://stackoverflow.com/questions/9427163/remove-duplicate-dict-in-list-in-python
-    result_unique = [i for n, i in enumerate(result) if i not in result[n+1:]]
+    result_unique = [i for n, i in enumerate(result) if i not in result[n + 1 :]]
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:304]: Done

File: pipreqs/pipreqs.py (hunk 1 above)
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:513]: What should we do in this case? Even if I solved this problem manually fixing this formater weirdness, next time someone runs the formatter it will just go back to this weird thing...

File: setup.py (hunk 2 above)
Comment (Fernando-crz) [reply setup.py:21]: Here's a reference on how it works: https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies

File: setup.py (hunk 3 above)
Comment (Fernando-crz) [reply setup.py]: Done @alan-barzilay

File: pipreqs/pipreqs.py [hunk 8]
```diff
@@ -128,6 +147,8 @@ def get_all_imports(
                     continue
                 else:
                     logging.error("Failed on file: %s" % file_name)
+                    if filter_ext(file_name, [".ipynb"]):
+                        logging.error("Magic command without % might be failed")
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:158]: Basically, jupyter notebooks have this feature called _Magic commands_, which give new possible commands to our program using IPython support. Magic commands can be called using the "%" sign, but there's a way to call those commands without using this specific symbol (check for `%automagic` in https://ipython.readthedocs.io/en/stable/interactive/magics.html). This log error basically says that **maybe** the program failed on the file because of that magic command conversion. As far as I know, those magic commands **probably** won't make our program fail, because we are only looking for imports with the AST tree.

**Comment by Fernando-crz** [comment]
Just fixed installing jupyter notebooks as an optional dependency. Basically, the program detects if the user has the library `nbconvert` installed, and if it is installed, the program will start detecting jupyter notebooks automatically when generating the _requirements.txt_ file.

I was thinking of maybe turning the detection of jupyter notebooks when creating the _requirements.txt_ file optional, using an optional argument with the name "`--ignore-notebooks`". What do you think? @alan-barzilay

File: pipreqs/pipreqs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (alan-barzilay) [reply pipreqs/pipreqs.py:199]: @Fernando-crz then why are we passing None at all? why not always default to the python standar that is utf-8? I think a new commit to change this pipreqs wide could be good idea

File: pipreqs/pipreqs.py (hunk 1 above)
Comment (alan-barzilay) [reply pipreqs/pipreqs.py]: @Fernando-crz first thing would be to understand why is it doing this haha are we sure this happens in the latest version of black? what are the exact options set? can anyone replicate this weird formatting?

File: pipreqs/pipreqs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (alan-barzilay) [reply pipreqs/pipreqs.py:158]: please add a test to check if magic commands without % would really make pipreqs to fail. If it is not an issue we can drop this if

File: pipreqs/pipreqs.py (hunk 1 above)
Comment (alan-barzilay) [reply pipreqs/pipreqs.py]: 23.9.1^ output of black version with flag for 120 character lines

File: setup.py (hunk 2 above)
Comment (alan-barzilay) [reply setup.py:21]: please add instructions to read me about how to deal with these optional dependencies

Comment (alan-barzilay) [reply setup.py:21]: also, if possible please reverse the behaviour so that by default the jupyter stuff is added and it is only excluded as a dependency when explicitly told to

**Comment by alan-barzilay** [comment]
> I was thinking of maybe turning the detection of jupyter notebooks when creating the _requirements.txt_ file optional, using an optional argument with the name "`--ignore-notebooks`". What do you think? @alan-barzilay

yeah, good idea!

**Review by alan-barzilay — CHANGES_REQUESTED** [review]
great work, we are almost there!

File: pipreqs/pipreqs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: Done

File: pipreqs/pipreqs.py [hunk 9]
```diff
@@ -488,6 +488,9 @@ def init(args):
     extra_ignore_dirs = args.get("--ignore")
     follow_links = not args.get("--no-follow-links")
     input_path = args["<path>"]
+
```
Comment (alan-barzilay) [inline pipreqs/pipreqs.py:491]: @Fernando-crz  woof woof

**Comment by Fernando-crz** [comment]
> I was thinking of maybe turning the detection of jupyter notebooks when creating the _requirements.txt_ file optional, using an optional argument with the name "`--ignore-notebooks`". What do you think? @alan-barzilay

Done.

File: pipreqs/pipreqs.py (hunk 9 above)
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: Done

File: pipreqs/pipreqs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: Just created a new file for magic commands inside the jupyter notebook test case (the file is named `magic_commands.ipynb`). As it turns out, tests are running successfully. It seems that magic commands with or without the percentage sign ("_%_") don't affect the ast search. 

With that in mind, i deleted the **if** related to this issue.

File: pipreqs/pipreqs.py (hunk 1 above)
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: I suggest that we can use darker instead (https://github.com/akaihola/darker/tree/master). It only runs the formatter over changes in the commits, meaning that, if I change this formatting back to it's origin state, it won't be changed in the next commits to the "next" branch.

By now, I will just revert the changes related to this specific line.

File: pipreqs/pipreqs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: Oddly enough, just noticed my tests ran successfully when run individually but failed when running with tox. I'll fix this asap

File: pipreqs/pipreqs.py [hunk 10]
```diff
@@ -145,11 +178,35 @@ def get_all_imports(path, encoding=None, extra_ignore_dirs=None, follow_links=Tr
     return list(packages - data)
```
Comment (mateuslatrova) [inline pipreqs/pipreqs.py:179]:
```suggestion

def get_file_extensions():
    return [".py", ".ipynb"] if PythonExporter and not ignore_notebooks else [".py"]
```

File: pipreqs/pipreqs.py [hunk 11]
```diff
@@ -102,13 +122,26 @@ def get_all_imports(path, encoding=None, extra_ignore_dirs=None, follow_links=Tr
         dirs[:] = [d for d in dirs if d not in ignore_dirs]
 
         candidates.append(os.path.basename(root))
-        files = [fn for fn in files if os.path.splitext(fn)[1] == ".py"]
+        if PythonExporter and not ignore_notebooks:
+            files = [fn for fn in files if filter_ext(fn, [".py", ".ipynb"])]
+        else:
+            files = [fn for fn in files if filter_ext(fn, [".py"])]
```
Comment (mateuslatrova) [inline pipreqs/pipreqs.py:128]:
```suggestion
            extensions =  get_file_extensions()
            files = [fn for fn in files if filter_ext(fn, extensions)]
```

File: pipreqs/pipreqs.py [hunk 12]
```diff
@@ -145,11 +178,35 @@ def get_all_imports(path, encoding=None, extra_ignore_dirs=None, follow_links=Tr
     return list(packages - data)
 
 
+def filter_ext(file_name, acceptable):
```
Comment (mateuslatrova) [inline pipreqs/pipreqs.py:181]: guess we could improve this functions name to something more meaningful.

something like `def file_extension_is_allowed(file_name, allow_list)`

File: pipreqs/pipreqs.py [hunk 13]
```diff
@@ -102,13 +122,26 @@ def get_all_imports(path, encoding=None, extra_ignore_dirs=None, follow_links=Tr
...
-        candidates += [os.path.splitext(fn)[0] for fn in files]
         for file_name in files:
             file_name = os.path.join(root, file_name)
-            with open(file_name, "r", encoding=encoding) as f:
-                contents = f.read()
+            contents = ""
+            if filter_ext(file_name, [".py"]):
+                with open(file_name, "r", encoding=encoding) as f:
+                    contents = f.read()
+            elif filter_ext(file_name, [".ipynb"]) and PythonExporter and not ignore_notebooks:
```
Comment (mateuslatrova) [inline pipreqs/pipreqs.py:143]: not the first time I see this condition `PythonExporter and not ignore_notebooks`.

maybe we could create a function with this condition so that its meaning is clearer. Example:

```
def notebooks_are_enabled():
   return PythonExporter and not ignore_notebooks
```

File: tests/test_pipreqs.py [hunk 14]
```diff
@@ -517,6 +519,48 @@ def test_output_requirements(self):
...
+        for item in imports:
+            self.assertTrue(item.lower() in self.modules, "Import is missing: " + item)
+        self.assertFalse("time" in imports)
+        self.assertFalse("logging" in imports)
+        self.assertFalse("curses" in imports)
+        self.assertFalse("__future__" in imports)
+        self.assertFalse("django" in imports)
+        self.assertFalse("models" in imports)
+        self.assertFalse("FastAPI" in imports)
+        self.assertFalse("sklearn" in imports)
```
Comment (mateuslatrova) [inline tests/test_pipreqs.py:537]:
```suggestion
        not_desired_imports = ["time", "logging", "curses", "__future__", "django", "models", "FastAPI", "sklearn"]        
        self.assertTrue(all(import not in imports for import in not_desired_imports))
```

File: tests/test_pipreqs.py [hunk 15]
```diff
@@ -559,6 +603,28 @@ def test_parse_requirements_handles_file_not_found(self, exit_mock):
...
+                "--pypi-server": None,
+                "--print": False,
+                "--diff": None,
+                "--clean": None,
+                "--mode": None,
+                "--ignore-notebooks": True,
+            }
+        )
+        assert os.path.exists(self.requirements_notebook_path) == 1
+        assert os.path.getsize(self.requirements_notebook_path) <= 1
```
Comment (mateuslatrova) [inline tests/test_pipreqs.py:626]: what does this check exactly?

File: pipreqs/pipreqs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: Just fixed the issue. As it turns out, tox was not considering the jupyter optional dependency when running it's tests, thus making my local test diverge from tox tests. Just added a line in the `tox.ini` doc to consider the optional dependency when testing. @alan-barzilay

File: setup.py (hunk 2 above)
Comment (Fernando-crz) [reply setup.py:21]: > also, if possible please reverse the behaviour so that by default the jupyter stuff is added and it is only excluded as a dependency when explicitly told to

It seems like there's no support for doing such thing. Check https://stackoverflow.com/questions/76376575/python-setuptools-exclude-dependencies-when-installing

What do you suggest doing in this case? Drop optional support? Maybe keep it as it is?

Comment (Fernando-crz) [reply setup.py:21]: > What do you suggest doing in this case? Drop optional support? Maybe keep it as it is?

@alan-barzilay

File: tests/test_pipreqs.py (hunk 14 above)
Comment (Fernando-crz) [reply tests/test_pipreqs.py:537]: Done

File: pipreqs/pipreqs.py (hunk 13 above)
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:143]: Done

File: pipreqs/pipreqs.py (hunk 12 above)
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:181]: Changed it to `file_ext_is_allowed` to avoid having too large names for functions

File: pipreqs/pipreqs.py (hunk 10 above)
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:179]: Decided to not implement this function because it makes the file extensions used by the program "hidden". The user will only know the file extension being used in the program if the user reads the function directly, making our code a little bit less readable

File: pipreqs/pipreqs.py (hunk 11 above)
Comment (Fernando-crz) [reply pipreqs/pipreqs.py:128]: > Decided to not implement this function because it makes the file extensions used by the program "hidden". The user will only know the file extension being used in the program if the user reads the function directly, making our code a little bit less readable

^^

File: tests/test_pipreqs.py (hunk 15 above)
Comment (Fernando-crz) [reply tests/test_pipreqs.py:616]: It just checks if the `requirements.txt` file generated by this test is created and is empty.

File: setup.py (hunk 2 above)
Comment (alan-barzilay) [reply setup.py:21]: hi, sorry for the delay. If there is no support for this feature, there is no support for it, what can we do haha
But i liked that `pip install --no-deps my_package` workaround, if we only import nbconverter/ipython when we use them (aka when we don't pass the cli option to ignore notebooks), someone that does this should be able to tailor their one dependencies without too much trouble. Maybe we should add a mention to this into the readme as well

Comment (alan-barzilay) [reply setup.py:21]: and, maybe I missed it but where do we use ipython directly? I found the import for nbvconvert, is it that ipython is a dependency of nbconvert? if thats the case, cant we omit it from the file and let nbconvert install its dependencies itself?

File: tests/test_pipreqs.py [hunk 16]
```diff
@@ -517,6 +515,42 @@ def test_output_requirements(self):
...
+
+    def test_ipynb_2_py(self):
+        """
+        Test the function ipynb_2_py() which converts .ipynb file to .py format
+        """
+        expected = pipreqs.get_all_imports(self.compatible_files["original"])
+        parsed = pipreqs.get_all_imports(self.compatible_files["notebook"])
+        self.assertEqual(expected, parsed)
+
+        parsed = pipreqs.get_all_imports(self.compatible_files["notebook"], encoding="utf-8")
```
Comment (alan-barzilay) [inline tests/test_pipreqs.py:543]: if the encoding is already utf-8, why are you adding it here? or is it from the original PR? if so, please update it accordingly

File: pipreqs/pipreqs.py [hunk 17]
```diff
@@ -48,9 +49,19 @@
 from yarg import json2package
 from yarg.exceptions import HTTPError
 
+try:
+    PythonExporter = None
+    ignore_notebooks = False
+    from nbconvert import PythonExporter
+except ImportError:
+    pass
+
```
[the remark this answers is not in the document]
Comment (alan-barzilay) [reply pipreqs/pipreqs.py:58]: oh, was the idea for it to be inside the exception handling? or are you making the existence of Python Exporter take precedence before the ignore notebooks flag?

**Review by alan-barzilay — CHANGES_REQUESTED** [review]
good work as always, dont forget to start thinking about how to adapt this PR to support poetry

File: .github/workflows/tests.yml [hunk 18]
```diff
@@ -39,7 +39,8 @@ jobs:
             - name: Install dependencies
               run: |
                   python -m pip install --upgrade pip
-                  pip install coverage docopt yarg requests
+                  pip install coverage docopt yarg requests nbconvert
+                  pip install .[jupyter]
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply .github/workflows/tests.yml:43]: "`.[jupyter]`" means that we are installing a library located in the current working directory and we are installing the optional depencies specified by the "`jupyter`" version. In this case, it means that we are installing _pipreqs_ with support for jupyter notebooks.

File: pipreqs/pipreqs.py [hunk 19]
```diff
@@ -145,11 +178,39 @@ def get_all_imports(path, encoding=None, extra_ignore_dirs=None, follow_links=Tr
     return list(packages - data)
 
 
+def notebooks_are_enabled():
+   return PythonExporter and not ignore_notebooks
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: woofn't

File: pipreqs/pipreqs.py [hunk 20]
```diff
@@ -256,7 +317,7 @@ def get_import_local(imports, encoding=None):
     # had to use second method instead of the previous one,
     # because we have a list in the 'exports' field
     # https://stackoverflow.com/questions/9427163/remove-duplicate-dict-in-list-in-python
-    result_unique = [i for n, i in enumerate(result) if i not in result[n + 1:]]
+    result_unique = [i for n, i in enumerate(result) if i not in result[n + 1 :]]
```
[the remark this answers is not in the document]
Comment (Fernando-crz) [reply pipreqs/pipreqs.py]: done

File: setup.py (hunk 2 above)
Comment (alan-barzilay) [reply setup.py:21]: please test if we really need the ipython dependency for  magics translation

File: tests/test_pipreqs.py (hunk 16 above)
Comment (alan-barzilay) [reply tests/test_pipreqs.py:543]: make this its one commit

**Comment by alan-barzilay** [comment]
Just so we have a written record of what we've just discussed:  let's drop the idea of optional dependencies.   Since we want the default option to include support for jupiter notebook, having optional dependencies don't  bring us any value.

 Besides that we also discussed the idea to further abstract some of the code processing steps to make them more auto contained and hide some of the logic with abstractions.  We'll also hide the jupiter logic behind the global variable ignore notebooks, if it is set to true nbconverter will never be imported.

@Fernando-crz  did i forget to mention anything?

**Comment by alan-barzilay** [comment]
@Fernando-crz can we close this PR if it is being continued in #414 ?

**Comment by Fernando-crz** [comment]
> Just so we have a written record of what we've just discussed: let's drop the idea of optional dependencies. Since we want the default option to include support for jupiter notebook, having optional dependencies don't bring us any value.
> 
> Besides that we also discussed the idea to further abstract some of the code processing steps to make them more auto contained and hide some of the logic with abstractions. We'll also hide the jupiter logic behind the global variable ignore notebooks, if it is set to true nbconverter will never be imported.
> 
> @Fernando-crz did i forget to mention anything?

LGMT

**Comment by Fernando-crz** [comment]
> @Fernando-crz can we close this PR if it is being continued in #414 ?

Sounds good to me.
